### PR TITLE
LOK-2503 : Create an GRPC and GraphQL interface to query Monitored Entities

### PR DESCRIPTION
**Pull request by Junaid-Khan-Nant**
## Description
Create a new GRPC service in inventory exporting interface to query Monitored Entities:
    Query all Monitored Entities
    Query Monitored Entities by provider
Add filters to query by location.
Create a GraphQL interface which forwards requests to the new GRPC service.

## Jira link(s)
https://opennms.atlassian.net/browse/LOK-2503

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java [hunk 1]
```diff
@@ -34,46 +34,7 @@
...
-import org.opennms.horizon.inventory.dto.SearchIpInterfaceQuery;
-import org.opennms.horizon.inventory.dto.SnmpInterfaceDTO;
-import org.opennms.horizon.inventory.dto.TagCreateListDTO;
-import org.opennms.horizon.inventory.dto.TagEntityIdDTO;
-import org.opennms.horizon.inventory.dto.TagListDTO;
-import org.opennms.horizon.inventory.dto.TagListParamsDTO;
-import org.opennms.horizon.inventory.dto.TagNameQuery;
-import org.opennms.horizon.inventory.dto.TagRemoveListDTO;
-import org.opennms.horizon.inventory.dto.TagServiceGrpc;
+import org.opennms.horizon.inventory.dto.*;
```
Comment (smunir-onms) [inline rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java:37]: avoid wildcard imports

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java [hunk 2]
```diff
@@ -93,4 +96,94 @@ public List<MonitoredEntity> getMonitoredEntities(final String tenantId, final l
                         .build())
                 .toList();
     }
+
+    @Override
+    public List<MonitorEntityResponse> getAllMonitoredEntities(String tenantId) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java:101]: The whole method is just a copy of already existing code...

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java [hunk 3]
```diff
@@ -93,4 +96,94 @@ public List<MonitoredEntity> getMonitoredEntities(final String tenantId, final l
...
+                                        yield null;
+                                    }
+                                })
+                        .build())
+                .map(entityMapper::modelToDTO)
+                .toList();
+    }
+
+    @Override
+    public List<MonitorEntityResponse> getMonitorEntitiesBySearchTerm(String tenantId, String searchTerm) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java:147]: Same

**Review by fooker — CHANGES_REQUESTED** [review]
This does not use the existing monitored entity structure at all. Make sure to not create own instances but use existing services.

**Review by fooker — CHANGES_REQUESTED** [review]
Please separate monitored entity status API and monitored entity search API.
Also the endpoint for getting all monitored entities for an tenant is missing.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+@RequiredArgsConstructor
+public class MonitoredEntitySearchGrpcService
+        extends MonitoredEntitySearchServiceGrpc.MonitoredEntitySearchServiceImplBase {
+
+    private final TenantLookup tenantLookup;
+    private final MonitoredEntityService monitoredEntityService;
+    private final MonitorEntityMapper mapper;
+
+    @Override
+    public void listAllMonitors(Int64Value request, StreamObserver<ListMonitorEntities> responseObserver) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java:52]: rename to `getAllMonitoredEntitiesByLocation`

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        responseObserver.onNext(ListMonitorEntities.newBuilder()
+                .addAllListMonitorEntityResponse(list)
+                .build());
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void getAllMonitoredEntitiesByTenant(Empty request, StreamObserver<ListMonitorEntities> responseObserver) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java:90]: rename to `getAllMonitoredEntities`. Everything is by tenant.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityProvider.java [hunk 6]
```diff
@@ -33,4 +33,6 @@ public interface MonitoredEntityProvider {
     String getProviderId();
 
     List<MonitoredEntity> getMonitoredEntities(final String tenantId, final long locationId);
+
+    List<MonitoredEntity> getMonitoredEntitiesByTenant(final String tenantId);
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityProvider.java:37]: I'm strongly against having such a method. Let the `MonitoredEntityService` figure all existing locations for a tenant and then call `getMonitoredEntities` for each of these locations.

File: shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+message SearchQuery {
+  optional int64  locationId = 1;
+  optional string providerId = 2;
+}
+
+service MonitoredEntitySearchService {
+  rpc listAllMonitors(google.protobuf.Int64Value) returns(ListMonitorEntities) {};
+  rpc getAllMonitoredEntitiesBySearchTerm(SearchQuery) returns(ListMonitorEntities) {};
+  rpc getAllMonitoredEntitiesByTenant(google.protobuf.Empty) returns(ListMonitorEntities){};
```
Comment (fooker) [inline shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto:58]: Rename here, too

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java [hunk 8]
```diff
@@ -44,6 +44,10 @@ public Stream<MonitoredEntity> getAllMonitoredEntities(final String tenantId, fi
                 .flatMap(provider -> provider.getMonitoredEntities(tenantId, locationId).stream());
     }
 
+    public Stream<MonitoredEntity> getAllMonitoredEntitesByTenant(final String tenantId) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java:47]: Rename here, too

File: shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto [hunk 9]
```diff
@@ -29,6 +29,8 @@ syntax = "proto3";
 
 import "google/protobuf/wrappers.proto";
 import "monitoredService.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/any.proto";
```
Comment (fooker) [inline shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto:33]: Revert

File: shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto [hunk 10]
```diff
@@ -41,7 +43,6 @@ message MonitoredEntityStateDTO {
   bool state = 4;
   int64 first_observation_time = 5;
 }
-
```
Comment (fooker) [inline shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto:44]: Revert.

**Review by fooker — CHANGES_REQUESTED** [review]

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+                .toList();
+
+        if (list.isEmpty()) {
+            Status status = Status.newBuilder()
+                    .setCode(Code.NOT_FOUND_VALUE)
+                    .setMessage("Monitored Entity not found")
+                    .build();
+            responseObserver.onError(StatusProto.toStatusRuntimeException(status));
+            return;
+        }
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java:82]: This is an outlier to the other behavior. Just return the empty list if no result could be found.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java [hunk 12]
```diff
@@ -35,15 +36,22 @@
...
 
     public Stream<MonitoredEntity> getAllMonitoredEntities(final String tenantId, final long locationId) {
         return this.providers.stream()
                 .flatMap(provider -> provider.getMonitoredEntities(tenantId, locationId).stream());
     }
 
+    public Stream<MonitoredEntity> getAllMonitoredEntities(final String tenantId) {
+        return monitoringLocationRepository.findByTenantId(tenantId).stream()
+                .flatMap(location -> this.providers.stream()
+                        .flatMap(provider -> provider.getMonitoredEntities(tenantId, location.getId()).stream()));
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java:52]: You can optimize by swapping:

```suggestion
        return this.providers.stream()
            .flatMap(provider -> monitoringLocationRepository.findByTenantId(tenantId).stream()
                .map(location -> provider.getMonitoredEntities(tenantId, location.getId()));
```

File: rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java [hunk 13]
```diff
@@ -50,4 +56,33 @@ public Mono<MonitoredEntityState> getMonitoredEntityState(
         var monitoredEntityState = mapper.protoToModel(monitoredEntityStateProto);
         return Mono.just(monitoredEntityState);
     }
+
+    @GraphQLQuery(name = "findAllMonitoredEntitiesByLocation")
+    public Flux<MonitoredDTO> findAllMonitoredEntitiesByLocation(
+            @GraphQLArgument(name = "locationId") long locationId, @GraphQLEnvironment ResolutionEnvironment env) {
+        return Flux.fromIterable(
```
Comment (fooker) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java:63]: There is a `Flux.fromStream` which can save you the `toList`

File: shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+message ListMonitorEntities {
+  repeated MonitorEntityResponse listMonitorEntityResponse = 1;
+}
+
+message SearchQuery {
+  optional int64  locationId = 1;
+  optional string providerId = 2;
+}
+
+service MonitoredEntitySearchService {
```
Comment (fooker) [inline shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto:55]:
```suggestion
service MonitoredEntityService {
```

**Review by fooker — CHANGES_REQUESTED** [review]
This also requires tests

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoredEntityStatusGrpcService.java [hunk 15]
```diff
@@ -21,24 +21,30 @@
...
 import org.springframework.stereotype.Component;
 
 @Component
 @RequiredArgsConstructor
 public class MonitoredEntityStatusGrpcService extends MonitoredEntityServiceGrpc.MonitoredEntityServiceImplBase {
 
+    public static final String EMPTY_TENANT_ID_MSG = "Tenant Id can't be empty";
     private final MonitoredEntityStateService monitorStatusService;
     private final TenantLookup tenantLookup;
+    private final DiscoveryMonitoredEntityProvider discoveryMonitoredEntityProvider;
```
[the remark this answers is not in the document]
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoredEntityStatusGrpcService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoredEntityStatusGrpcService.java [hunk 16]
```diff
@@ -57,6 +63,47 @@ public void getMonitoredEntityState(StringValue request, StreamObserver<Monitore
...
+        tenantIdOptional.ifPresentOrElse(
+                tenantId -> {
+                    try {
+                        var list = discoveryMonitoredEntityProvider.getMonitorEntitiesBySearchTerm(
+                                tenantId, request.getValue());
+                        responseObserver.onNext(ListMonitorEntities.newBuilder()
+                                .addAllListMonitorEntityResponse(list)
+                                .build());
+                        responseObserver.onCompleted();
+                    } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoredEntityStatusGrpcService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java (hunk 2 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java]: removed.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java (hunk 3 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/model/DiscoveryMonitoredEntityProvider.java]: removed.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java (hunk 4 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java (hunk 5 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityProvider.java (hunk 6 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityProvider.java]: removed.

File: shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto (hunk 7 above)
Comment (Junaid-Khan-Nant) [reply shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java (hunk 8 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java]: Done.

File: shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto (hunk 9 above)
Comment (Junaid-Khan-Nant) [reply shared-lib/inventory/src/main/proto/monitoredServiceStatus.proto]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java (hunk 11 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/discovery/MonitoredEntitySearchGrpcService.java]: removed.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java (hunk 12 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java]: Done.

File: rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java (hunk 13 above)
Comment (Junaid-Khan-Nant) [reply rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java]: Done.

File: shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto (hunk 14 above)
Comment (Junaid-Khan-Nant) [reply shared-lib/inventory/src/main/proto/monitoredEntitySearch.proto]: Done.

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.opennms.horizon.server.model.inventory;
+
+import com.google.protobuf.Any;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class MonitoredDTO {
```
Comment (fooker) [inline rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java:30]: Please rename to `MonitoredEntityDTO`

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    public String entityId;
+    public long locationId;
+    public String type;
+    public Any config;
```
Comment (fooker) [inline rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java:34]: `Any` is not supported in a GraphQL response. You have multiple routes to choose from here:
- Serialize the config to JSON and make the field a string
- Go all in and figure out how to represent arbitrary scalar data in GraphQL (maybe a `JsonNode` could work)
- Create a follow up issue which we can defere until we know what UI needs there

**Review by fooker — CHANGES_REQUESTED** [review]

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java (hunk 18 above)
Comment (smunir-onms) [reply rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java]: > `Any` is not supported in a GraphQL response. You have multiple routes to choose from here:
> 
> * Serialize the config to JSON and make the field a string
> * Go all in and figure out how to represent arbitrary scalar data in GraphQL (maybe a `JsonNode` could work)
> * Create a follow up issue which we can defere until we know what UI needs there
Issue created: https://opennms.atlassian.net/browse/LOK-2545

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java (hunk 17 above)
Comment (Junaid-Khan-Nant) [reply rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredDTO.java]: Done

File: rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java [hunk 19]
```diff
@@ -39,7 +38,7 @@
 import reactor.core.publisher.Mono;
 
 @RequiredArgsConstructor
-@GraphQLApi
+// @GraphQLApi
```
Comment (fooker) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GraphQLMonitoredEntityService.java:41]: Remove

File: inventory/main/src/test/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    private MonitoringLocation location;
+    private MonitoringLocationRepository locationRepository;
+
+    @BeforeEach
+    public void setUp() {
+
+        locationRepository = mock(MonitoringLocationRepository.class);
+        TaskSetPublisher taskSetPublisher = mock(TaskSetPublisher.class);
+
+        MonitoredEntityProvider provider1 = mock(MonitoredEntityProvider.class);
```
Comment (fooker) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityServiceTest.java:60]: I think you are better of creating a real impl of the `MonitoredEntityProvider` interface in the test. Let it serve some static data that you can test against.

**Review by fooker — APPROVED** [review]

**Pull request merged**